src/lib/SaferNames/Name.hs

```diff
@@ -0,0 +1,478 @@
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+unitName :: Name UnitScope
+unitName = UnsafeMakeName "TheName"
+
+unitScope :: Scope UnitScope
+unitScope = UnsafeMakeEnv $ LM.singleton unitRawName ()
+  where UnsafeMakeName unitRawName = unitName
+
+envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
+envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
+  where f' rawName = f $ UnsafeMakeName rawName
+
+envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
+envPairs (UnsafeMakeEnv m) = do
+  let (names, vals) = unzip $ LM.assocs m
+  (unsafeMakePlainBinderList names, vals)
+
+unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
+unsafeMakePlainBinderList names = case names of
+  [] -> unsafeCoerceB Empty
+  (name:rest) -> Nest (UnsafeMakeBinder name) $ unsafeMakePlainBinderList rest
+
+-- === specialized traversals ===
+
+type NameSet n = S.Set (Name n)
+
+fmapNames :: HasNamesE e => Scope b -> (Name a -> Name b) -> e a -> e b
+fmapNames scope f e = runIdentity $ traverseNamesE scope t e
+  where t = newNameTraversal (pure . f)
+
+foldMapNames :: (HasNamesE e, Monoid a) => (Name n -> a) -> e n -> a
+foldMapNames f e =
+  execWriter $ traverseNamesE unitScope t e
+  where t = newNameTraversal \name -> tell (f name) >> return unitName
+
+freeNames :: HasNamesE e => e n -> NameSet n
+freeNames e = foldMapNames S.singleton e
+
+-- should be equivalent to `\ext -> fmapNames (injectName ext)`
+injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
+injectNamesL _ = unsafeCoerceE
+
+projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
+projectNamesL scope scopeFragment e = traverseNamesE scope t e
+  where t = newNameTraversal \name -> case projectName scopeFragment name of
+                                        Left name' -> Just name'
+                                        Right _    -> Nothing
+
+-- === common scoping patterns ===
+
+data Abs (binder::B) (body::E) (n::S) where
+  Abs :: binder n l -> body l -> Abs binder body n
+
+data FreshAbs (binder::B) (body::E) (n::S) where
+  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
+
+data NestPair (b1::B) (b2::B) (n::S) (l::S) where
+  NestPair :: b1 n l -> b2 l l' -> NestPair b1 b2 n l'
+
+data Nest (binder::B) (n::S) (l::S) where
+  Nest  :: binder n h -> Nest binder h l -> Nest binder n l
+  Empty ::                                  Nest binder n n
+
+type PlainBinderList = Nest PlainBinder
+
+infixl 7 :>
+data AnnBinderP (b::B) (ann ::E) (n::S) (l::S) =
+  (:>) (b n l) (ann n)
+  deriving (Show)
+
+infixl 7 :>>
+data AnnBinderListP (b::B) (ann::E) (n::S) (l::S) =
+  (:>>) (Nest b n l) [ann n]
+
+-- === environment variants ===
+
+type RenameEnv i o = Env i (Name o)
+
+-- covariant in `o`, contravariant in `i`.
+newtype EnvE (e::E) (i::S) (o::S) = EnvE { fromEnvE :: Env i (e o) }
+
+-- covariant in `n`
+newtype RecEnv (e::E) (n::S) = RecEnv { fromRecEnv :: Env n (e n) }
```

Comment:
It's recursive because the payloads have the names in scope. For example:

```
-- nonrecursive let
data Let n l where Let :: Binder n l -> Expr n -> Let n l

--recursive let
data LetRec n l where LetRec :: Binder n l -> Expr l -> Let n l
```